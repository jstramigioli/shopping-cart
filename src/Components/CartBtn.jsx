import cartImg from '../assets/cart.svg'
import style from './CartBtn.module.css'
import backImg from '../assets/back.svg' 
import PropTypes from 'prop-types'

function CartBtn({onClick, showingCart, cart}) {

    return (
        <button onClick={onClick} 
        className={style.btn + ' ' + (showingCart ? style.flip : '')}>
        
            <img src={cartImg} className={style.cart} alt="cart" />
            <img src={backImg} className={style.back} alt="back" />
            {
                (cart.reduce((acc, current)=> acc + current.quantity, 0) > 0 && showingCart == false)&& <p className={style.quantity}>{cart.reduce((acc, current)=> acc + current.quantity, 0)}</p>
            }
            
        </button>
    )
}

CartBtn.propTypes = {
    onClick: PropTypes.func,
    showingCart: PropTypes.bool,
    cart: PropTypes.array
}

export default CartBtn