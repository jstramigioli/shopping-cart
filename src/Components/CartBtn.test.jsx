import { render, screen } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import userEvent from "@testing-library/user-event";
import CartBtn from "./CartBtn";
import styles from "./CartBtn.module.css";

describe('CartBtn', () => {
    it('renders button correctly', () => {
        const mockCart = [{ id: 1, quantity: 2 }];
        render( <CartBtn onClick={() => {}} showingCart={false} cart={mockCart} />)
        expect(screen.getByRole("button")).toBeInTheDocument();
        expect(screen.getByAltText("cart")).toBeInTheDocument();
        expect(screen.getByAltText("back")).toBeInTheDocument();    
    })
    
    it('runs onClick when button pressed', async () => {
        const user = userEvent.setup();
        const mockOnClick = vi.fn(); 

        render(
            <CartBtn onClick={mockOnClick} showingCart={false} cart={[]} />
          );

          await user.click(screen.getByRole("button"));

          expect(mockOnClick).toHaveBeenCalledOnce()
    })

    it('shows quantity only if cart showingCart is false', () => {
        const mockCart = [{quantity:1}, {quantity:3}]
        const {rerender} = render(
            <CartBtn onClick={()=>{}} showingCart={false} cart={mockCart} />
          );
        expect(screen.getByText('4')).toBeInTheDocument()
        rerender(
            <CartBtn onClick={() => {}} showingCart={true} cart={mockCart} />
          );
          expect(screen.queryByText("4")).not.toBeInTheDocument();
    })

    it('doesnt show quantity if cart is empty', () => {
        const mockCart = []
        render(<CartBtn onClick={()=>{}} showingCart={false} cart={mockCart} />)
        expect(screen.queryByText(/[0-9]+/)).not.toBeInTheDocument();
    })
    
    it('applies flip classname when showingCart is true', () => {
        const { container } = render(
            <CartBtn onClick={() => {}} showingCart={true} cart={[]} />
          );
        const button = container.querySelector("button");
        expect(button).toHaveClass(styles.flip); 
    })

    it('matches snapshot', () => {
        const mockCart = [{quantity:1}, {quantity:3}]
        const { container } = render(
            <CartBtn onClick={() => {}} showingCart={false} cart={mockCart} />
          );
        expect(container).toMatchSnapshot()
    })
})

