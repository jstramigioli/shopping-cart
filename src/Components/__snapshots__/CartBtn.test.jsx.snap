// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartBtn > matches snapshot 1`] = `
<div>
  <button
    class="_btn_b818be "
  >
    <img
      alt="cart"
      class="_cart_b818be"
      src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20width='48'%20height='48'%20viewBox='0%200%2024%2024'%3e%3cpath%20fill='currentColor'%20d='M0%201h4.764l.545%202h18.078l-3.666%2011H7.78l-.5%202H22v2H4.72l1.246-4.989L3.236%203H0V1Zm7.764%2011h10.515l2.334-7H5.855l1.909%207ZM4%2021a2%202%200%201%201%204%200a2%202%200%200%201-4%200Zm14%200a2%202%200%201%201%204%200a2%202%200%200%201-4%200Z'/%3e%3c/svg%3e"
    />
    <img
      alt="back"
      class="_back_b818be"
      src="data:image/svg+xml,%3c?xml%20version='1.0'%20encoding='UTF-8'%20standalone='no'?%3e%3csvg%20width='48px'%20height='48px'%20viewBox='0%200%2018%2020'%20version='1.1'%20xmlns='http://www.w3.org/2000/svg'%20xmlns:xlink='http://www.w3.org/1999/xlink'%3e%3c!--%20Generator:%20Sketch%2041.2%20(35397)%20-%20http://www.bohemiancoding.com/sketch%20--%3e%3ctitle%3eic_back%3c/title%3e%3cdesc%3eCreated%20with%20Sketch.%3c/desc%3e%3cdefs%3e%3c/defs%3e%3cg%20id='Icons'%20stroke='none'%20stroke-width='1'%20fill='none'%20fill-rule='evenodd'%20stroke-linecap='round'%20stroke-linejoin='round'%3e%3cg%20id='24-px-Icons'%20transform='translate(-219.000000,%20-26.000000)'%20stroke='%23000000'%3e%3cg%20id='ic_back'%20transform='translate(216.000000,%2024.000000)'%3e%3cg%20transform='translate(12.000000,%2012.000000)%20scale(-1,%201)%20translate(-12.000000,%20-12.000000)%20'%20id='forward'%3e%3cg%20transform='translate(4.000000,%203.000000)'%20stroke-width='2'%3e%3cpath%20d='M0,9%20L16,9'%20id='Line'%3e%3c/path%3e%3cpath%20d='M16,9%20L7.93774223,0.937742233'%20id='Line'%3e%3c/path%3e%3cpath%20d='M16,9%20L7.93774223,17.0622578'%20id='Line'%3e%3c/path%3e%3c/g%3e%3c/g%3e%3c/g%3e%3c/g%3e%3c/g%3e%3c/svg%3e"
    />
    <p
      class="_quantity_b818be"
    >
      4
    </p>
  </button>
</div>
`;
